### DISCUSSION TITLE
users/me API not working anymore

### DISCUSSION THREAD

**Discussion by kkahara** [directus/directus#6510] [Archive]
/users/me  API worked fine about a week ago, but I am getting an error message now.

The error message:

```json
{
  "errors": [
    {
      "message": "Invalid user credentials.",
      "extensions": {
        "code": "INVALID_CREDENTIALS"
      }
    }
  ]
```

I am passing 

Authorization: "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZC..."

in the header. 

Is this a bug?

**Comment by paescuj** [comment]
Can you please provide more information? What's your environment, browser, Directus version, ...?
How are you using the API? Please note that tokens have an expiry date and you need to refresh them or issue a new one: https://docs.directus.io/reference/api/authentication/

**Comment by kkahara** [comment]
Thank you. I have tried to call the API from my React.JS app and a REST client tool but got the same error. 

Here is my React code:

```js
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [userToken, setUserToken] = useState("");
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [message, setMessage] = useState("");

  async function handleSubmit(event) {
    event.preventDefault();

    axios
      .post(
        "https://directus.isaiahinstitute.jp/auth/login",
        {
          email: email,
          password: password,
        },
        {
          headers: {
            "content-type": "application/json",
          },
        }
      )
      .then((response) => {
        // Handle success.
        localStorage.setItem("userToken", response.data.access_token);
        localStorage.setItem("userIsLoggedIn", true);
        setIsLoggedIn(true);
        setUserToken(response.data.data.access_token);
        setMessage("successfully obtained the token");
      })
      .catch((error) => {
        // Handle error.
        console.log("An error occurred:", error.response);
        setMessage(
          "確認できませんでした。ユーザー名かパスワードをお確かめください。"
        );
        if (error.response.statusCode === "400") {
          setMessage("確認できませんでした。");
        }
        setIsLoggedIn(false);
      });

    axios
      .get("https://directus.isaiahinstitute.jp/users/me", {
        Authorization: "Bearer " + userToken,
      })
      .then((response) => {
        // Handle success.
        setMessage("ようこそ、" + response.data.first_name + "さん");
        localStorage.setItem("userEmail", response.data.data.email);
        localStorage.setItem("userID", response.data.data.id);
        setTimeout(() => browserHisotry.goBack(), 1000);
      })
      .catch((error) => {
        // Handle error.
        console.log("An error occurred:", error.response);
        if (error.response.statusCode === "400") {
          setMessage("ユーザー情報を入手できませんでした。");
        }
        setIsLoggedIn(false);
      });
  }
```

Below is a screenshot of my REST client app:

![image](https://user-images.githubusercontent.com/47410653/123504714-08cace00-d618-11eb-97d2-4ce5ab098d3b.png)

I can successfully obtain the access_token but don't know the next step--to obtain the user info by using the access_token. I don't think I can use /users/:id API because I don't have the id yet. I also don't know what permission I need to give to the role of the user to obtain the info. I use Edge browser and Directus version 9.

![image](https://user-images.githubusercontent.com/47410653/123504818-b807a500-d618-11eb-83e8-961337156605.png)


Thank you

**Comment by paescuj** [comment]
As this is not a issue with Directus itself but a question, can you please close this issue and open up a thread under https://github.com/directus/directus/discussions.
Just mention me there and I will try to help you!

**Comment by paescuj** [comment]
@kkahara
Some tips:
* Requests to "users/me" must be issued *after* user has been logged in and a valid token is present. (axios is async)
* Tokens will expire after some time and must be refreshed. (see docs)
* Permission depends on your use case, but read access to fields like first name, last name is granted by default.
* You may want to use the SDK. (https://docs.directus.io/reference/sdk/)

**Reply by rijkvanzanten** [reply]
To visualize that:

```js
axios.post('https://directus.isaiahinstitute.jp/auth/login', { email, password })
  .then((authResponse) => {
    const token = authResponse;
    
    axios.get('https://directus.isaiahinstitute.jp/users/me', { headers: { Authorization: `Bearer ${token}` }})
      .then((meResponse) => {
        console.log(meResponse.data.data);
      });
  })
```

As you can see, this nested promises approach gets unwieldy pretty fast. I would recommend using `async/await` in this case:

```js
async function loginAndGetUser() {
  const authResponse = await axios.post('https://directus.isaiahinstitute.jp/auth/login', { email, password });
  const token = authResponse.data.data.access_token;
  const meResponse = await axios.get('https://directus.isaiahinstitute.jp/users/me', { headers: { Authorization: `Bearer ${token}` }})
  console.log(meResponse.data.data);
}

loginAndGetUser();
```

---

As @paescuj suggested, it might be easier to use `@directus/sdk` in this case. Here's the same function, but using the JS SDK:

```js
import { Directus } from '@directus/sdk';

const directus = new Directus('https://directus.isaiahinstitute.jp/');

async function loginAndGetUser() {
  await directus.auth.login({ email, password });
  const me = await directus.users.me.read()
  console.log(me);
}

loginAndGetUser();
```

**Reply by rijkvanzanten** [reply]
The reason it doesn't work in your original code is because the two axios methods are called at the exact same time @kkahara:

![CleanShot 2021-06-26 at 11 58 51@2x](https://user-images.githubusercontent.com/9141017/123518878-f1bac900-d675-11eb-9b98-039c7f3e4212.png)

Number 2 doesn't wait for number 1 to be done, so when number 2 fires, `userToken` is undefined 👍🏻

**Reply by kkahara** [reply]
Thank you so much for pointing out axios method was async and needed to nest the methods and recommending me to use the SDK. I will try the SDK.
